modules/function/src/main/java/com/opengamma/strata/function/calculation/deposit/TermDepositCalculationFunction.java

```diff
@@ -0,0 +1,129 @@
+
+  /**
+   * The calculations by measure.
+   */
+  private static final ImmutableMap<Measure, SingleMeasureCalculation> CALCULATORS =
+      ImmutableMap.<Measure, SingleMeasureCalculation>builder()
+          .put(Measure.PAR_RATE, TermDepositMeasureCalculations::parRate)
+          .put(Measure.PAR_SPREAD, TermDepositMeasureCalculations::parSpread)
+          .put(Measure.PRESENT_VALUE, TermDepositMeasureCalculations::presentValue)
+          .put(Measure.PV01, TermDepositMeasureCalculations::pv01)
+          .put(Measure.BUCKETED_PV01, TermDepositMeasureCalculations::bucketedPv01)
+          .build();
+
+  /**
+   * Creates an instance.
+   */
+  public TermDepositCalculationFunction() {
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public Set<Measure> supportedMeasures() {
+    return CALCULATORS.keySet();
+  }
+
+  @Override
+  public Optional<Currency> defaultReportingCurrency(TermDepositTrade target) {
+    return Optional.of(target.getProduct().getCurrency());
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public FunctionRequirements requirements(TermDepositTrade trade, Set<Measure> measures) {
+    TermDeposit product = trade.getProduct();
+
+    Set<DiscountCurveKey> discountCurveKeys =
+        ImmutableSet.of(DiscountCurveKey.of(product.getCurrency()));
+
+    return FunctionRequirements.builder()
+        .singleValueRequirements(discountCurveKeys)
+        .timeSeriesRequirements()
+        .outputCurrencies(product.getCurrency())
+        .build();
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public Map<Measure, Result<ScenarioResult<?>>> calculate(
+      TermDepositTrade trade,
+      Set<Measure> measures,
+      CalculationMarketData scenarioMarketData) {
+
+    // expand the trade once for all measures and all scenarios
+    ExpandedTermDeposit product = trade.getProduct().expand();
+
+    // loop around measures, calculating all scenarios for one measure
+    Map<Measure, Result<ScenarioResult<?>>> results = new HashMap<>();
+    for (Measure measure : measures) {
+      results.put(measure, calculate(measure, trade, product, scenarioMarketData));
+    }
+    return results;
+  }
+
+  // calculate one measure
+  private Result<ScenarioResult<?>> calculate(
+      Measure measure,
+      TermDepositTrade trade,
+      ExpandedTermDeposit product,
+      CalculationMarketData scenarioMarketData) {
+
+    SingleMeasureCalculation calculator = CALCULATORS.get(measure);
+    if (calculator == null) {
+      return Result.failure(FailureReason.INVALID_INPUT, "Unsupported measure: {}", measure);
+    }
+    return Result.of(() -> calculator.calculate(trade, product, scenarioMarketData));
+  }
+
+  //-------------------------------------------------------------------------
+  @FunctionalInterface
+  interface SingleMeasureCalculation {
```

Comment:
Can we abstract this out rather than having to define roughly the same interface in every function?